File: questApplication/src/main/java/ru/nekit/android/qls/quest/answer/common/QuestAnswerChecker.java
package ru.nekit.android.qls.quest.answer.common;

import android.support.annotation.NonNull;

import ru.nekit.android.qls.quest.common.Quest;

public class QuestAnswerChecker<T> implements IAnswerChecker<T> {

    @Override
    public boolean checkAlternativeInput(@NonNull Quest quest, @NonNull T answer) {
        return quest.getAnswer().equals(answer);
    }

    @Override
    public boolean checkStringInputFormat(@NonNull Quest quest, @NonNull String value) {
        boolean result = false;
        Class answerClass = quest.getAnswerClass();
        try {
            if (answerClass == Integer.class) {
                Integer.parseInt(value);
                result = true;
            } else if (answerClass == Double.class) {
                Double.parseDouble(value);
                result = true;
            }
        } catch (NumberFormatException ignored) {
        }
        if (answerClass == String.class) {
            result = true;
        }
        return result;
    }

    @Override
    public boolean checkStringInput(@NonNull Quest quest, @NonNull String value) {
        boolean result = false;
        Class answerClass = quest.getAnswerClass();
        if (answerClass == Integer.class) {
            result = quest.getAnswer() == Integer.valueOf(value);
        } else if (answerClass == Double.class) {
            result = quest.getAnswer() == Double.valueOf(value);
        } else if (answerClass == String.class) {
            result = ((String) quest.getAnswer()).toLowerCase().equals(value.toLowerCase());
        }
        return result;
    }
}

File: questDomain/src/main/java/ru/nekit/android/qls/domain/answerChecker/common/QuestAnswerChecker.kt
package ru.nekit.android.qls.domain.answerChecker.common

import ru.nekit.android.qls.domain.model.quest.Quest

open class QuestAnswerChecker<in R : Quest> : IAnswerChecker<R> {

    override fun checkAlternativeInput(quest: R, answer: Any): Boolean {
        return quest.answer == answer
    }

    override fun checkStringInputFormat(quest: R, value: String): Boolean {
        var result = false
        val answerClass = quest.answerClass
        try {
            if (answerClass == Int::class.java) {
                Integer.parseInt(value)
                result = true
            } else if (answerClass == Double::class.java) {
                java.lang.Double.parseDouble(value)
                result = true
            }
        } catch (ignored: NumberFormatException) {
        }
        if (answerClass == String::class.java) {
            result = true
        }
        return result
    }

    override fun checkStringInput(quest: R, value: String): Boolean {
        val answerClass = quest.answerClass
        return when (answerClass) {
            Int::class.java -> quest.answer == Integer.valueOf(value)
            Double::class.java -> quest.answer == java.lang.Double.valueOf(value)
            String::class.java -> (quest.answer as String).toLowerCase() == value.toLowerCase()
            else -> false
        }
    }
}